### ISSUE TITLE
If the raytracer doesn't start, please post in here

### ISSUE THREAD

**Issue by 20k** [20k/geodesic_raytracing#1]
Hi there! Lots of people have bundled over here from twitter which is nice

If you run into any problems getting this to run, please post: Your OS, GPU, and the contents of the cmd prompt window that starts up. I haven't gotten around to implementing disk logging, which in hindsight I probably should do

Currently its only been tested on windows/AMD, though Nvidia *should* work. Linux may work through wine, though a native build is planned. I'd be interested if it works or not there!

**Comment by jpmvferreira** [comment]
Hello, opening `RelativityWorkshop.exe` crashes immediately with no information whatsoever, either on `debug.txt` or on the terminal, and opening `swt.exe` shows the first menu and pressing `Create New` crashes with no information as well.
Specs:
- OS: Windows 10
- GPU: Intel +1050Ti (laptop)

I prefer using Linux over Windows, so if there is some makefile for my to run on Linux I would give it a shot and see what happens. It would probably be simpler to debug, even though the laptop is the same, because I have the dedicated GPU disabled on Linux.

**Comment by 20k** [comment]
That's unfortunate to hear. Out of curiosity, do you have steam installed on your laptop? swt.exe is a steam workshop tool, but hey at least it starts

There's an `xmake` build system, but there's no linux configuration as I don't have a testing environment. In theory, it should largely just compile and run under linux, but there'd be some porting work involved

[build_public.zip](https://github.com/user-attachments/files/18390732/build_public.zip)

I've uploaded a more heavily instrumented version that will hopefully tell me where the problem lies, if you're willing to do a bit of debugging. It should output at least *something* on the CLI, and then everything else will hopefully be in debug.txt

**Comment by jpmvferreira** [comment]
> Out of curiosity, do you have steam installed on your laptop?
Yes, but having it running on the background or not doesn't matter, it still crashes.

Here's the contents of `debug.txt` after running the executable from the file explorer:
```
Parsed args
Set up steam context
Workshop started
Workshop callback
Main menu
Pre render window
```
Running `cmd /K RelativityWorkshop.exe` on the CMD prompt showed nothing on the CLI.

I wouldn't be too worried about my system in particular, as laptops with Intel+NVIDIA are known to cause weird issues. Besides, this GPU is already a little bit old and maybe the newer GPUs don't have this issue.
I will try to run this program in a laptop that doesn't have this dual GPU sometime this week and see what happens.

**Comment by 20k** [comment]
Ok thank you for testing it! It looks like there's something odd going on with context creation, and I'd guess that its something to do with OpenCL being set up if I had to guess. Its probable that its something to do with the mixed laptop hardware - I've tested this on some very old hardware so as long as its opencl 1.2 compatible it should be fine. Out of curiosity if you give it another try, forcing it to run on the dedicated hardware via the control panel might help

[build_public.zip](https://github.com/user-attachments/files/18419536/build_public.zip)

This is another version with some more debugging built in if you happen to decide to give it another shot, I'm hoping that its just picking the wrong platform for OpenCL

**Comment by jpmvferreira** [comment]
Thanks for the build, running it from the terminal I get the following written in the `debug.txt`:
```
Parsed args
Set up steam context
Workshop started
Workshop callback
Main menu
Pre render window
Making backend
Glfw make backend
Set vsync
Pre make OpenCL context
Start context
Platform found: Intel(R) OpenCL HD Graphics
Platform found: NVIDIA CUDA
Selecting: NVIDIA CUDA
Get device IDs
Post device IDs
Supports gl sharing
Pre-make context
Built context
```

But, I found a solution!
I thought changing the settings in the NVIDIA control panel wouldn't do anything because it's already selecting the NVIDIA GPU, but there might be some sort of conflict and if I force this to be the case on the NVIDIA control panel, then it works.
If I had to guess, maybe the program is asking for the dedicated GPU and NVIDIA is telling it "nah" for some reason, so it crashes.
Either way, problem solved.

**Comment by jpmvferreira** [comment]
This is clearly off-topic, but I was having fun playing around with it a thought to myself: wouldn't this be better with a joystick? The answer is, yes, it is! See below for a small video.
If you want to replicate this, get yourself a joystick, you might need to use [x360](https://www.x360ce.com/) to have the joystick properly configured and recognized, and use [joytokey](https://joytokey.net/en/) to bind keys to the joystick.
All you have to do is bind WASD and the arrow keys to the left/right stick and off you go, into your cosmic journey!

I would like to say, for information sake, that on my 1050ti mobile, I get a not very good performance on raytracing Kerr when looking straight into the BH (it reminded me of valve engine when the game would run at hundreds of fps when you weren't looking at anything and then have 15 fps because somebody decided to throw smoke everywhere).

https://github.com/user-attachments/assets/d8c7bc74-4bad-4df2-8116-927e57eb6e6a
